src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs

```diff
@@ -29,66 +30,113 @@ internal abstract partial class AbstractAddParameterCheckCodeRefactoringProvider
+        protected override async Task<ImmutableArray<CodeAction>> GetRefactoringsForAllParametersAsync(
+            Document document, SyntaxNode functionDeclaration, IMethodSymbol methodSymbol,
+            IBlockOperation blockStatementOpt, IEnumerable<SyntaxNode> listOfParameterNodes, int position, CancellationToken cancellationToken)
         {
-            // Only should provide null-checks for reference types and nullable types.
-            if (!parameter.Type.IsReferenceType &&
-                !parameter.Type.IsNullable())
+
+            if (listOfParameterNodes == null)
             {
                 return ImmutableArray<CodeAction>.Empty;
             }
 
-            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
-            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            // List to keep track of the valid parameters
+            var listOfParametersOrdinals = new List<int>();
 
-            // Look for an existing "if (p == null)" statement, or "p ?? throw" check.  If we already
-            // have one, we don't want to offer to generate a new null check.
-            //
-            // Note: we only check the top level statements of the block.  I think that's sufficient
-            // as this will catch the 90% case, while not being that bad an experience even when 
-            // people do strange things in their constructors.
-            if (blockStatementOpt != null)
+            foreach (var parameterNode in listOfParameterNodes)
             {
-                if (!CanOffer(blockStatementOpt.Syntax))
+                var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                var parameter = (IParameterSymbol)semanticModel.GetDeclaredSymbol(parameterNode, cancellationToken);
+                if (ParameterValidForNullCheck(document, parameter, semanticModel, blockStatementOpt, cancellationToken))
                 {
-                    return ImmutableArray<CodeAction>.Empty;
+                    listOfParametersOrdinals.Add(parameter.Ordinal);
                 }
+            }
 
-                foreach (var statement in blockStatementOpt.Operations)
+            if (!listOfParametersOrdinals.Any())
+            {
+                return ImmutableArray<CodeAction>.Empty;
+            }
+
+            // Great.  The list has parameters that need null checks. Offer to add null checks for all.
+            return ImmutableArray.Create<CodeAction>(new MyCodeAction(
+                 FeaturesResources.Add_null_checks_for_all_parameters,
+                     c => DocumentRetriever(document, functionDeclaration, blockStatementOpt, listOfParametersOrdinals, c)));
+
+            // DocumentRetriever updates all necessary variables per each parameter and calls for AddNullCheckAsync
+            async Task<Document> DocumentRetriever(
```

Comment:
please don't make this a local function. it's far too easy to unintentionally capture outer state and make it last to long.

Reply:
static local function?

Reply:
why was this resolved?